// @botpoison/node/dist/index.js
"use strict";
var tslib_1 = require("tslib");
var Botpoison = /** @class */ (function () {
    function Botpoison(_a) {
        var secretKey = _a.secretKey;
        this.secretKey = secretKey;
    }
    Botpoison.prototype.verify = function (solution) {
        return (0, tslib_1.__awaiter)(this, void 0, void 0, function () {
            var _this = this;
            return (0, tslib_1.__generator)(this, function (_a) {
                return [2 /*return*/, new Promise(function (resolve, reject) {
                        var https = require("https"); // eslint-disable-line @typescript-eslint/no-var-requires
                        var body = JSON.stringify({
                            secretKey: _this.secretKey,
                            solution: solution,
                        });
                        var options = {
                            hostname: "api.botpoison.com",
                            path: "/verify",
                            method: "POST",
                            headers: {
                                "Content-Type": "application/json",
                                "Content-Length": body.length,
                            },
                        };
                        var request = https.request(options, function (response) {
                            response.on("data", function (data) {
                                if (response.statusCode === 200) {
                                    resolve(JSON.parse(data.toString()));
                                }
                                else {
                                    reject(JSON.parse(data.toString()));
                                }
                            });
                        });
                        request.on("error", function (error) {
                            reject(error);
                        });
                        request.write(body);
                        request.end();
                    })];
            });
        });
    };
    return Botpoison;
}());
module.exports = Botpoison;
